java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -21,336 +21,312 @@
-   * 
-   * @param relation another relation whose left operand is the same as the receiver's right operand
-   * @return the resulting relation or null if the conjunction of both relations does not bring any new information
-   */
-  @CheckForNull
-  protected abstract BinaryRelation combinedAfter(BinaryRelation relation);
+  private static BinaryRelation equalityTransitiveBuilder(BinaryRelation a, BinaryRelation b) {
+    BinaryRelation equality = isEqualityRelation(a.kind) ? a : b;
+    BinaryRelation other = equality == a ? b : a;
+    SymbolicValue transitiveOperand = equality.differentOperand(other);
+    boolean leftTransitive = equality.hasOperand(other.leftOp);
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithEqual(EqualRelation relation);
+    if (equality.kind == METHOD_EQUALS && other.kind == EQUAL) {
+      return equalityTransitiveBuilder(other, equality);
+    }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithNotEqual(NotEqualRelation relation);
+    BinaryRelation deduced = binaryRelation(other.kind,
+      leftTransitive ? transitiveOperand : other.leftOp,
+      leftTransitive ? other.rightOp : transitiveOperand);
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithMethodEquals(MethodEqualsRelation relation);
-
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithNotMethodEquals(NotMethodEqualsRelation relation);
+    // FIXME remove this sanity check
+    Preconditions.checkState(!a.equals(deduced), "Deduced relation " + deduced + " is equal to input relation " + a);
+    Preconditions.checkState(!b.equals(deduced), "Deduced relation " + deduced + " is equal to input relation " + b);
+    return deduced;
+  }
 
   @CheckForNull
-  protected abstract BinaryRelation combinedWithGreaterThan(GreaterThanRelation relation);
+  private static BinaryRelation lessThanTransitiveBuilder(BinaryRelation lessThan, BinaryRelation other) {
+    BinaryRelation deduced = null;
+    if (other.kind == LESS_THAN) {
+      // a < x && x < b -> a < b
+      if (lessThan.rightOp.equals(other.leftOp)) {
+        deduced = binaryRelation(LESS_THAN, lessThan.leftOp, other.rightOp);
+      }
+      // x < a && b < x -> b < a
+      if (lessThan.leftOp.equals(other.rightOp)) {
+        deduced = binaryRelation(LESS_THAN, other.leftOp, lessThan.rightOp);
+      }
+    }
+    if (other.kind == GREATER_THAN_OR_EQUAL) {
+      // a < x && b >= x -> a < b
+      if (lessThan.rightOp.equals(other.rightOp)) {
+        deduced = binaryRelation(LESS_THAN, lessThan.leftOp, other.leftOp);
+      }
+      // x < a && x >= b -> b < a
+      if (lessThan.leftOp.equals(other.leftOp)) {
+        deduced = binaryRelation(LESS_THAN, other.rightOp, lessThan.rightOp);
+      }
+    }
 
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithGreaterThanOrEqual(GreaterThanOrEqualRelation relation);
+    // FIXME remove this sanity checks
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)